### DISCUSSION TITLE
Generic questions about the challenge

### DISCUSSION THREAD

**Discussion by danpak94** [FETS-AI/Challenge#40] [Q&A] [answered]
Hello,

Thank you for organizing this nice challenge. I have a few generic questions about Task_1.

1) Can we find out how many teams are registered/participating for this challenge?

2) Is this the only forum for asking questions? (I want to make sure I'm not missing other sources of information)

3) I noticed that most of the heavy lifting seems to be performed in openfl and gandlf. I have a lot of questions about openfl, and I was wondering if we are expected to ask questions about them here or if we are expected to find out those answers ourselves through the openfl documentation. The two questions below might give you an sense of why I am interested in learning more about the openfl framework.

3b) How can I cut down the wall time spent per round? Running the run_challenge_experiment was very slow, even when I set collaborator training selection to just 1 collaborator at all rounds. For example, if I have some problem with my setup (e.g. not enough disk space to save checkpoints), I don't want to have to wait 48 hours before I find that out, and testing that wouldn't require having to train and validate with all collaborators.

3c) How do I plot the metrics displayed by collaborator.py (in openfl) during training? Can we only visualize the results after the entire thing has been trained? Again, this was problematic because training even the simplest test takes a very long time.

Thank you

**Comment by sarthakpati** [comment]
Hi @danpak94,

Thank you for your interest in the FeTS Challenge! 

1. We cannot give this information out at this time, unfortunately. In case this changes, we will inform participants on a public forum so that everyone receives the information together.

2. This is the preferred forum to ask questions. The other method is [via email](mailto:challenge@fets.ai), if you don't wish to make your query publicly visible.

3a. You can ask questions related to OpenFL/GaNDLF here itself, and the organizers will either answer directly or provide references to the documentation, whenever appropriate.

3b & c: tagging @brandon-edwards @msheller @psfoley @alexey-gruzdev 

Cheers,
Sarthak on behalf of the FeTS Challenge Organizing Committee

**Reply by danpak94** [reply]
Thanks for the clarifications!

**Comment by brandon-edwards** [comment, marked as answer]
Hi danpak94,
Thanks for your questions. Please find responses to 3b and 3c below.

> Hello,
> 
> Thank you for organizing this nice challenge. I have a few generic questions about Task_1.
> 
> 1. Can we find out how many teams are registered/participating for this challenge?
> 2. Is this the only forum for asking questions? (I want to make sure I'm not missing other sources of information)
> 3. I noticed that most of the heavy lifting seems to be performed in openfl and gandlf. I have a lot of questions about openfl, and I was wondering if we are expected to ask questions about them here or if we are expected to find out those answers ourselves through the openfl documentation. The two questions below might give you an sense of why I am interested in learning more about the openfl framework.
> 
> 3b) How can I cut down the wall time spent per round? Running the run_challenge_experiment was very slow, even when I set collaborator training selection to just 1 collaborator at all rounds. For example, if I have some problem with my setup (e.g. not enough disk space to save checkpoints), I don't want to have to wait 48 hours before I find that out, and testing that wouldn't require having to train and validate with all collaborators.

For testing purposes (ie, something like making sure you have enough disk space to save the checkpoints), you can set the value of 'challenge_metrics_validation_interval' to a large value in order to skip the validation step for most rounds (if your test does not require seeing much model validation). Limiting to one collaborator is also a good idea, and when doing so try to pick a smaller one. You can also make your own partitioning csv, in order to create institutions of whatever size you wish for  such tests. Running on GPU, I get through a single round this way (single small institution) in minutes.

In addition, evaluating Hausdorff distance is computationally expensive and so adds quit a bit to the runtime. To address this (for those that do not care to collect this aspect of validation), we will soon be pushing a change that allows the removal of Hausdorff from validation altogether.
> 
> 3c) How do I plot the metrics displayed by collaborator.py (in openfl) during training? Can we only visualize the results after the entire thing has been trained? Again, this was problematic because training even the simplest test takes a very long time.

I would suggest using checkpointing, then running your experiment with progressively greater values of rounds_to_train (like a run with 5 while enabling saving of a checkpoint, then run by restoring from that checkpoint with a new value of rounds_to_train of 10, ...). Each time you get a dataframe as the return of run_challenge_experiment, and you can plot those as they come in. A short script wrapping run_challenge_experiment could automate this.
> 
> Thank you

**Reply by danpak94** [reply]
Thanks for the detailed response. I understand what you are saying and they all make sense, but I am having trouble producing the desired behavior by changing the parameters and files you mentioned.

I tried creating my own csv by changing the partition ID to 00 for the first two entries of partitioning_1.csv (i.e. changed to FeTS21_Training_001,00 and FeTS21_Training_096,00). I changed my choose_training_collaborators function to always output ['00']. But it seems like these only control which data are used for training - when I run the run_challenge_experiment function, I see all collaborators (00 through 17) running the "aggregated_model_validation" task. Is this the expected behavior, or did I miss something? I also set rounds_to_train=1 and challenge_metrics_validation_interval=100, so I was expecting no model validation to run, but it seems like that's not really the case, given that the aggregated_model_validation task kept running for all the collaborators.

Thanks for the clarification and tip on checkpointing. I will try it out once I get to that point.

**Reply by brandon-edwards** [reply]
Sorry for the confusion, by 'make your own partitioning csv' I also meant that you can delete data rows in the csv, thus leaving only one or two very small institutions.

> Thanks for the detailed response. I understand what you are saying and they all make sense, but I am having trouble producing the desired behavior by changing the parameters and files you mentioned.
> 
> I tried creating my own csv by changing the partition ID to 00 for the first two entries of partitioning_1.csv (i.e. changed to FeTS21_Training_001,00 and FeTS21_Training_096,00). I changed my choose_training_collaborators function to always output ['00']. But it seems like these only control which data are used for training - when I run the run_challenge_experiment function, I see all collaborators (00 through 17) running the "aggregated_model_validation" task. Is this the expected behavior, or did I miss something?

That is the expected behavior for the changes you made. That will reduce the training time considerably but keep the validation time the same. Instead, try doing what you did in addition to deleting all remaining rows (i.e., keep only institution '00'). This will define a federation with only this one small collaborator.

>  I also set rounds_to_train=1 and challenge_metrics_validation_interval=100, so I was expecting no model validation to run, but it seems like that's not really the case, given that the aggregated_model_validation task kept running for all the collaborators.

This is also expected, but simply due to the fact that validation is always run at round 0 and 1. Validation is run whenever the condition
round % challenge_metrics_validation_interval == 0
is met. So in the case you mention if you increased the rounds to train to 101, validation would run at round 0, 1, and 100.

> 
> Thanks for the clarification and tip on checkpointing. I will try it out once I get to that point.

**Reply by danpak94** [reply]
Thanks for the clarification. I was able to indeed speed up the process by deleting most of the rows in the csv file.

However, when I run it (with everything set up as we discussed here), I run into an error,
![image](https://user-images.githubusercontent.com/21160618/123101086-aef0bb00-d401-11eb-88da-1341a136a004.png)
which, when I checked, was because tensor_db.tensor_db does not have performance_evaluation_metric_binary_DICE_WT (I think it's only a problem for the chosen training collaborator). But I was not sure how I should proceed from there.

**Reply by brandon-edwards** [reply]
Could you provide all of the argument values passed to  run_challenge_experiment, and let me know if this error occurs within round 0 (or specify what round this occurs in). Also, did you reduce the csv to include only one collaborator? And are you sure that the collaborator selection function returns this collaborator?

**Reply by danpak94** [reply]
`scores_dataframe = run_challenge_experiment(
    aggregation_function=weighted_average_aggregation,
    choose_training_collaborators=one_fixed_collaborator,
    training_hyper_parameters_for_round=constant_hyper_parameters,
    validation_functions=[('sensitivity', sensitivity), ('specificity', specificity)],
    institution_split_csv_filename= '/home/dhp26/code_data/Challenge/Task_1/openfl-workspace/fets_challenge_workspace/partitioning_dp.csv',
    brats_training_data_parent_dir='/home/dhp26/project/data/fets_data/MICCAI_FeTS2021_TrainingData',
    db_store_rounds=1,
    rounds_to_train=1,
    device='cuda',
    challenge_metrics_validation_interval=3,
    save_checkpoints=True,
    restore_from_checkpoint_folder = None)`

`def one_fixed_collaborator(collaborators,
                            db_iterator,
                            fl_round,
                            collaborators_chosen_each_round,
                            collaborator_times_per_round):
    return ['00']`

The error occurs within round 0.
The csv contains two collaborators - 0 and 17. 
![image](https://user-images.githubusercontent.com/21160618/123135727-1fa6d000-d420-11eb-8bc7-5b6e805b4997.png)
I'm pretty sure the selection function is returning ['00'] because I see the logging info from collaborator.py saying "17 received the following tasks: ['aggregated_model_validation'] and several more lines, and then "00 received the following tasks: ['aggregated_model_validation', 'train', 'locally_tuned_model_validation']

**Reply by brandon-edwards** [reply]
I don't see anything that could be a problem @psfoley or @alexey-gruzdev can you advise?

**Reply by Linardos** [reply]
I had the same issue in the Notebook. I pasted the relevant parts of the code in a script but the error persists (I use the small_split.csv): 
![image](https://user-images.githubusercontent.com/26694607/123641160-a1389c80-d82a-11eb-9d93-0208d7c317a4.png)

**Reply by Linardos** [reply]
Apparently, the same happens even if I use the complete partition_1.csv just took longer for it to crash:
![image](https://user-images.githubusercontent.com/26694607/123682062-351e5e80-d853-11eb-90ba-e821d76622dd.png)

**Reply by Vishruth-Sham** [reply]
@Linardos, did you get a fix for this problem? I am facing the same issue.

**Reply by Linardos** [reply]
Nope, still no solution

**Reply by sarthakpati** [reply]
Hi @Linardos @Vishruth-Sham, we are investigating this and will get back to you ASAP.

**Reply by brandon-edwards** [reply]
@Vishruth-Sham @Linardos we apologize for the delay in finding this solution, but setting db_store_rounds to a value greater than 1 should resolve the issue. Please let us know if this is not the case on your end.

**Reply by Vishruth-Sham** [reply]
@brandon-edwards This fixed the issue. Thanks
